src/pybind/mgr/dashboard_v2/restresource.py

```diff
@@ -0,0 +1,95 @@
+import json
+
+import cherrypy
+
+
+def _takes_json(func):
+    def inner(*args, **kwargs):
+        body = cherrypy.request.body.read()
+        try:
+            data = json.loads(body.decode('utf-8'))
+        except json.JSONDecodeError as e:
+            raise cherrypy.HTTPError(400, 'Failed to decode JSON: {}'.format(str(e)))
+        return func(data, *args, **kwargs)
+    return inner
+
+
+def _returns_json(func):
+    def inner(*args, **kwargs):
+        cherrypy.serving.response.headers['Content-Type'] = "application/json"
+        ret = func(*args, **kwargs)
+        return json.dumps(ret).encode('utf8')
+    return inner
+
+def json_error_page(status, message, traceback, version):
+    return json.dumps(dict(status=status, detail=message, traceback=traceback, version=version))
+
+
+class RESTResource(object):
+    """
+    Base class for providing a RESTful interface to a resource.
+
+    To use this class, simply derive a class from it and implement the methods
+    you want to support.  The list of possible methods are:
+
+    * list()
+    * bulk_set(data)
+    * create(data)
+    * bulk_delete()
+    * get(key)
+    * set(data, key)
+    * delete(key)
+
+    Test with curl:
+
+    curl -H "Content-Type: application/json" -X POST -d '{"username":"xyz","password":"xyz"}'  http://127.0.0.1:8080/foo
+    curl http://127.0.0.1:8080/foo
+    curl http://127.0.0.1:8080/foo/0
+
+    """
+
+    _cp_config = {
+                  'request.error_page': {'default': json_error_page},
+                  }
+
+    def _no_implemented(self, is_element):
```

Comment:
I think `_not_implemented` is more intuitive as a name here.

Reply:
fixed